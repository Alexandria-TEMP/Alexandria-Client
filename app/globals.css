@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
