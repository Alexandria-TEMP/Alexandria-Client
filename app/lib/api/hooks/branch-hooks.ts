"use client";

import useSWR, { SWRResponse } from "swr";
import { fetchBranchData } from "../services/branch-api";
import { BranchUnionT, idBranchUnionT } from "@/lib/types/branch-union";

/**
 * Fetches branch or closed branch data in a unified object
 * @returns
 *    data: branch and optionally closed branch data (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useBranchData(
  id: idBranchUnionT,
): SWRResponse<BranchUnionT, Error> {
  return useSWR(id, fetchBranchData, {
    revalidateOnMount: true,
    refreshInterval: 5000,
  });
}
