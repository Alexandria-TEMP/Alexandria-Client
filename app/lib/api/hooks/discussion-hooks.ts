import { DiscussionT, MemberT, idT } from "@/lib/types/api-types";
import { fetchDiscussionData } from "../services/discussion-api";
import useSWR, { SWRResponse } from "swr";
import { fetchMemberData } from "../services/member-api";

/**
 * Fetches discussion and author data
 * @returns
 *    data: discussion data and author (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useDiscussionAndAuthorData(
  id: idT,
): SWRResponse<{ discussion: DiscussionT; author: MemberT }, Error> {
  return useSWR({ id }, async ({ id }) => {
    const discussion = await fetchDiscussionData(id);
    const author = await fetchMemberData(discussion.memberID);
    return { discussion, author };
  });
}
