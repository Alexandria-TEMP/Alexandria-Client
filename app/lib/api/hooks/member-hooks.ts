"use client";

import useSWR, { SWRResponse } from "swr";
import { MemberT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Hook that fetches all members from the database, using SWR for caching, loading and error states
 * Fetching all members from the database sounds like a very bad idea, but we would need some kind of lazy loading for MultiSelectAutocomplete otherwise
 * that is currently in todo
 * NOTE: currently the endpoint actually only returns id, first and last names as that is all that is needed
 * @returns the member objects as array, as well as other SWR states (loading, error)
 */
export function useFetchMembers(): SWRResponse<MemberT[], Error> {
  return useSWR(baseUrl + "/members", async (...args) => {
    const response = await fetch(...args); //by default a get request
    await validateResponse(response);
    const members: MemberT[] = (await response.json()) as MemberT[];
    return members;
  });
}
