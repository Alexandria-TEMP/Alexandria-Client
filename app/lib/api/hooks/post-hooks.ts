"use client";

import { PostUnionT, idPostUnionT } from "@/lib/types/post-union";
import useSWR, { SWRResponse } from "swr";
import {
  fetchDataForPostOfUnkownType,
  fetchPostData,
} from "../services/post-api";
import { ScientificFieldTagT, idT } from "@/lib/types/api-types";
import {
  fetchScientificFieldContainer,
  fetchScientificFieldsFromContainer,
} from "../services/fields-api";
import { fetchDiscussionContainer } from "../services/discussion-api";

/**
 * Fetches post or project post data in a unified object
 * @param id the id of the post or project post whose data we are getting
 * @returns
 *    data: post and optionally project post data (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function usePostData(id: idPostUnionT): SWRResponse<PostUnionT, Error> {
  return useSWR(id, fetchPostData);
}

/**
 * Function that fetches all data about a post, including the list of scientific field tag ids
 * This is necessary to conditionally fetch scientific fields, after post data is looaded
 * @param id post/project post id
 * @returns
 *    data: scientific fields IDs, post, and optionally project post data (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function usePostAndScientificFieldData(
  id: idPostUnionT,
): SWRResponse<PostUnionT & { scientificFieldTagIDs: idT[] }, Error> {
  return useSWR(id, async (id) => {
    const postData = await fetchPostData(id);
    const scientificFieldsContainer = await fetchScientificFieldContainer(
      postData.post.scientificFieldTagContainerID,
    );
    return {
      ...postData,
      scientificFieldTagIDs: scientificFieldsContainer.scientificFieldTagIDs,
    };
  });
}

/**
 * Fetches post or project post data in a unified object in addition to
 * the post's list of scientificFields and how many discussions are associated
 * with it. Intended for use in PostPreviewCard.
 * @param id the id of the post or project post whose data we are getting
 * @returns
 *    data: post and optionally project post data (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function usePostPreviewData(id: idT): SWRResponse<
  PostUnionT & {
    scientificFields: ScientificFieldTagT[];
    numDiscussions: number;
  },
  Error
> {
  return useSWR({ id }, async ({ id }) => {
    const postUnion = await fetchDataForPostOfUnkownType(id);
    const scientificFields = await fetchScientificFieldsFromContainer(
      postUnion.post.scientificFieldTagContainerID,
    );
    const numDiscussions = (
      await fetchDiscussionContainer(postUnion.post.discussionContainerID)
    ).discussionIDs.length;

    return { ...postUnion, scientificFields, numDiscussions };
  });
}
