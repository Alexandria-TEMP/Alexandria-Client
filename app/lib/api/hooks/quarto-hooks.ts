"use client";

import useSWR, { SWRResponse } from "swr";
import { parseFileTree } from "../../file-tree-handler";
import { FileTreeT } from "../../types/file-tree";
import { QuartoContainerT } from "../../types/quarto-container";
import { baseUrl, validateResponse } from "../api-common";
import { useMemo } from "react";
import { toKebabCase } from "../string-utils";

/**
 * Builds URL path for quarto project API calls
 * @param container.id post or branch ID
 * @param container.type container type, ie if Quarto project is in a post or branch
 */
const buildResourcePath = ({ id, type }: QuartoContainerT) =>
  `${baseUrl}/${type === "branch" ? "branches" : "posts"}/${id}`;

/**
 * Fetches HTML render of a Quarto project.
 * @param container.id post or branch ID
 * @param container.type container type, ie if Quarto project is in a post or branch
 * @returns
 *    data: text contents of the HTML render (or undefined if not loaded),
 *    isPending: true if render is not finalized (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useRender(
  container: QuartoContainerT,
): SWRResponse<string, Error> & { isPending: boolean } {
  const swrResponse: SWRResponse<string, Error> = useSWR(
    `${buildResourcePath(container)}/render`,
    async (...args) => {
      const res = await fetch(...args);
      if (res.status === 202) return "pending";
      await validateResponse(res);
      return res.text();
    },
  );

  const isPending = useMemo(
    () => swrResponse.data === "pending",
    [swrResponse.data],
  );

  return {
    isPending,
    ...swrResponse,
  };
}

/**
 * Fetches file tree of a Quarto project and converts it to nested object format
 * @param container.id post or branch ID
 * @param container.type container type, ie if Quarto project is in a post or branch
 * @returns
 *    data: file tree (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useFileTree(
  container: QuartoContainerT,
): SWRResponse<FileTreeT, Error> {
  return useSWR(`${buildResourcePath(container)}/tree`, async (...args) => {
    const res = await fetch(...args);
    await validateResponse(res);
    const responseTree = (await res.json()) as { [key: string]: number };
    return parseFileTree(responseTree);
  });
}

/**
 * Fetches file contents of a given file in Quarto project
 * @param container.id post or branch ID
 * @param container.type container type, ie if Quarto project is in a post or branch
 * @param path path of the file within the project
 * @returns
 *    data: file contents as text (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useFileContents(
  container: QuartoContainerT,
  path: string,
): SWRResponse<string, Error> {
  return useSWR(
    `${buildResourcePath(container)}/file/${path}`,
    async (...args) => {
      const res = await fetch(...args);
      await validateResponse(res);
      return res.text();
    },
  );
}

/**
 * Downloads quarto project to client computer
 * @param container.id post or branch ID
 * @param container.type container type, ie if Quarto project is in a post or branch
 * @param filename name of the downloaded file, will be random if undefined
 */
export function downloadProject(
  container: QuartoContainerT,
  filename?: string,
) {
  fetch(`${buildResourcePath(container)}/repository`)
    .then(async (res) => {
      await validateResponse(res);
      return res.blob();
    })
    .then((blob) => {
      // Creates a link object to download file
      const link = document.createElement("a");
      link.href = window.URL.createObjectURL(blob);

      // Set name for download
      if (filename) link.download = toKebabCase(filename);

      // Download file
      link.click();
    })
    .catch((reason: Error) => {
      // Let user know if anything went wrong
      alert(`Failed to download files. \n[${reason.message}]`);
    });
}
