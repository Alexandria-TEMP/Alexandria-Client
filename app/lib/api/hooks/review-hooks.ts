import { BranchReviewT, idT } from "@/lib/types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import useSWR, { SWRResponse } from "swr";

/**
 * Fetches review data
 * @returns
 *    data: review data (or undefined if not loaded),
 *    error: error thrown by fetcher (or undefined),
 *    isLoading: if there's an ongoing request and no "loaded data"
 */
export function useReviewData(id: idT): SWRResponse<BranchReviewT, Error> {
  return useSWR(`${baseUrl}/branches/reviews/${id}`, async (url: string) => {
    const res = await fetch(url);
    await validateResponse(res);
    return (await res.json()) as BranchReviewT;
  });
}
