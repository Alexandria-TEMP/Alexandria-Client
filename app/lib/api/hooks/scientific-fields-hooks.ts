"use client";

import useSWR, { SWRResponse } from "swr";
import { ScientificFieldTagT, idT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import { fetchScientificFieldsFromContainer } from "../services/fields-api";

/**
 * Hook that fetches all scientific fields from the database, using SWR for caching, loading and error states
 * Fetching all tags from the database sounds like a very bad idea, but we would need some kind of lazy loading
 * for MultiSelectAutocomplete otherwise that is currently in todo
 * @returns the scientific field objects as array, as well as other SWR states (loading, error)
 */
export function useScientificFields(): SWRResponse<
  ScientificFieldTagT[],
  Error
> {
  return useSWR(baseUrl + "/tags/scientific", async (...args) => {
    const response = await fetch(...args);
    await validateResponse(response);
    const scientificFieldTags: ScientificFieldTagT[] =
      (await response.json()) as ScientificFieldTagT[];
    return scientificFieldTags;
  });
}

/**
 * Hook that gets all the scientific field tag objects in a container, by id
 * @param id of the container
 * @returns an SWR response containing an array of scientific field tag id, loading and erorr states
 */
export function useScientificFieldsByContainer(
  id: idT,
): SWRResponse<ScientificFieldTagT[], Error> {
  return useSWR({ cid: id }, (key) =>
    fetchScientificFieldsFromContainer(key.cid),
  );
}
