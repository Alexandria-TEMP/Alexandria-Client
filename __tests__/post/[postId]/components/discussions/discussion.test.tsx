import { expect, describe, it } from "@jest/globals";
import { render } from "@testing-library/react";
import Discussion from "@/post/[postId]/components/discussions/discussion";
import { dummyDiscussion, dummyMembers } from "~/__tests__/__utils__/dummys";
import { useDiscussionAndAuthorData } from "@/lib/api/hooks/discussion-hooks";

jest.mock("@/lib/api/hooks/discussion-hooks");

describe("Discussion", () => {
  (useDiscussionAndAuthorData as jest.Mock).mockReturnValue({
    data: { discussion: dummyDiscussion, author: dummyMembers[0] },
    isLoading: false,
  });

  it("matches snapshot", () => {
    const { container } = render(<Discussion id={dummyDiscussion.id} />);
    expect(container).toMatchSnapshot();
  });
});
