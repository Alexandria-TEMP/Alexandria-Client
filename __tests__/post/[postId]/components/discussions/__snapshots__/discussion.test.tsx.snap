// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discussion matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden height-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none"
    tabindex="-1"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      <div
        class="flex flex-row gap-x-2 items-baseline -mt-2"
      >
        <p
          class="font-semibold"
        >
          Marie Curie
        </p>
        <div
          class="grow"
        />
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-default-foreground data-[hover=true]:bg-default/40"
          type="button"
        >
          Reply
        </button>
      </div>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sunt ut ducimus amet ex qui eius corrupti reiciendis, quibusdam suscipit, aspernatur ipsum. Reprehenderit libero molestias nostrum eum sed? Illo, quidem ad.
      </p>
    </div>
  </div>
</div>
`;
